[FLINK-11813] Introduces basic JobManagerRunner implementation

XComp opened the pull request:
## What is the purpose of the change

Two goals are covered by this PR:
* Consolidation of the cleanup phase (moving cleanup logic into `GlobalCleanupStage`) and providing a `localCleanup` method in the `Dispatcher` with now retry mechanism involved, yet!
* Implements a basic `JobManagerRunner` class that only instantiates the checkpoint-related components and implements close logic for them. No retry mechanism is applied, yet.

Follow-ups:
* `RetryStategy` needs to be integrated
* The two classes implementing `JobManagerRunner` (i.e. `JobMasterServiceLeadershipRunner` and `CheckpointJobDataCleanupRunner`) have to be consolidated. General logic (the leader election) should be moved into an abstract class
* We might want to integrate the `ExecutionGraphInfoStore` in the `CheckpointJobDataCleanupRunner` as a fallback.

## Brief change log

See individual commits and their messages for descriptions.

## Verifying this change

[TODO] This PR contains only a prototype. Additional efforts (like implementing tests, adding documentation) need to be applied.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

XComp committed:
[FLINK-11813] Removes unused classloader parameter from CheckpointRecoveryFactory.createRecoveredCompletedCheckpointStore

XComp committed:
[FLINK-11813] Moves CheckpointRecoveryFactory initialization based on JobManager configuration into the interface

XComp committed:
[hotfix] Refactors nested if statements

XComp committed:
[hotfix] Minor fixes in ApplicationStatus

XComp committed:
[FLINK-11813] Renames ArchivedExecutionGraph.createFromInitializingJob into more generic ArchivedExecutionGraph.createSparseArchivedExecutionGraph

XComp committed:
[FLINK-11813] Adds CheckpointResourcesCleanupRunner

XComp committed:
[FLINK-11813] Simplifies code

dmvk commented on flink-core/src/main/java/org/apache/flink/core/fs/local/LocalFileSystem.java:
```diff
@@ -200,6 +200,9 @@ public boolean delete(final Path f, final boolean recursive) throws IOException
     private boolean delete(final File f) throws IOException {
+        if (!f.exists()) {
+            return true;
+        }
```
Do we have a failing test case for this? Ideally one that covers all filesystems (I'm not sure whether we have such abstract test in place)

XComp replied:
Yes, I still have to add that 👍

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:
```diff
@@ -804,76 +859,27 @@ private void registerJobManagerRunnerTerminationFuture(
-            } catch (IOException e) {
-                log.warn("Could not properly mark job {} result as clean.", jobId, e);
-            }
+    private void cleanUpJobResult(JobID jobId) {
```
`commitJobResult`?

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/CheckpointJobDataCleanupRunner.java:
```diff
@@ -0,0 +1,238 @@
+
+    private void initializeCompletedCheckpointStore() {
+        try {
+            this.completedCheckpointStore =
```
The cleanup executor uses a fixed thread pool [1], so we're not guaranteed to always have a same thread here.

[1] https://github.com/apache/flink/blob/release-1.14.0/flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java#L314

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/CheckpointJobDataCleanupRunner.java:
```diff
@@ -0,0 +1,238 @@
+    public CompletableFuture<Acknowledge> cancel(Time timeout) {
+        Preconditions.checkState(
+                resultFuture != null, "The CheckpointJobDataCleanupRunner was not started, yet.");
+        if (resultFuture.cancel(true)) {
```
What is the use-case for cancelling here? This will fail the future, but wouldn't the cleanup still keep executing?

XComp replied:
You're right. I have to look into it more closely.

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/CheckpointJobDataCleanupRunner.java:
```diff
@@ -0,0 +1,238 @@
+        return new ExecutionGraphInfo(
+                ArchivedExecutionGraph.createSparseArchivedExecutionGraph(
+                        jobResult.getJobId(),
+                        "we might want to move the job name into the job result",
```
do we also need this for the web ui?

XComp replied:
yes, that was my understanding. we need to provide some dummy response for now

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/CheckpointJobDataCleanupRunner.java:
```diff
@@ -0,0 +1,238 @@
+                    try {
+                        completedCheckpointStore.shutdown(getJobStatus(), checkpointsCleaner);
+                    } catch (Exception e) {
+                        exception = e;
```
would it make sense to simply fail right away here? If I understand that correctly we'll eventually retry anyway

XComp replied:
Thinking about it once more: You're probably right. For HA, it would be most likely be the same error anyway when accessing ZK or k8s

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/JobManagerMetricGroup.java:
```diff
@@ -84,7 +86,17 @@ public JobManagerJobMetricGroup addJob(JobID jobId, String jobName) {
+    }
+
+    private void cleanup(JobID jobId) {
         if (jobId == null) {
```
under what circumstances can the jobId be null?

dmvk commented on flink-core/src/main/java/org/apache/flink/util/concurrent/FutureUtils.java:
```diff
@@ -766,6 +772,22 @@ public RetryException(Throwable cause) {
+                            CompletableFuture<T>,
+                            Collection<CompletableFuture<T>>,
+                            FutureUtils.ConjunctFuture<Collection<T>>>
+                    combineAll() {
```
seems to be unused

XComp replied:
I will revert the FutureUtils changes again...

dmvk commented on flink-core/src/main/java/org/apache/flink/util/concurrent/FutureUtils.java:
```diff
@@ -938,6 +960,49 @@ public int getNumFuturesCompleted() {
+
+        private final Function<Collection<CompletableFuture<I>>, O> finisherFunction;
+        private static final Set<Characteristics> CHARACTERISTICS =
+                new HashSet<>(Arrays.asList(Characteristics.UNORDERED, Characteristics.CONCURRENT));
```
Is this really concurrent? Looking at the jdk collectors, concurrent one is for example `#toConcurrentMap`. This one is basically a list collector, which only uses `IDENTITY_FINISH` characteristic.

Also I'd say that this implementation is order preserving.

Overall, I have a feeling that collecting to list is good enough and this is not worth the troubles.

```
        return FutureUtils.completeAll(
                jobRelatedCleanups.stream()
                        .map(c -> c.globalCleanupAsync(jobId, cleanupExecutor))
                        .collect(Collectors.toList()));
```

If you want to go down this path, we should have a good test coverage for the new methods.

XComp replied:
Agree, it's probably not worth the effort 👍 Initially, I though it might be a useful easy extension.

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraph.java:
```diff
@@ -337,7 +337,7 @@ public static ArchivedExecutionGraph createFrom(
      * will be empty, only job status and error-related fields are set.
      */
-    public static ArchivedExecutionGraph createFromInitializingJob(
+    public static ArchivedExecutionGraph createSparseArchivedExecutionGraph(
```
Why do we need to rename this? It generates a lot of noise in the PR as i touches many files.

XComp replied:
The renaming became necessary because it would be used in different places (in initialization and cleanup phase for the job). Hence, the old name was wrong. 

I try to keep the commits separated logically in such big PRs to make it possible to go over the PR commit by commit. Is this something that helps or is it unnecessary effort?

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:
```diff
@@ -824,46 +826,61 @@ public boolean deletePermanent(JobID jobId, PermanentBlobKey key) {
+    @Override
+    public void localCleanup(JobID jobId) throws IOException {
         checkNotNull(jobId);
+        deleteLocalData(jobId);
```
Should we also obtain a RW lock here?

XComp replied:
Thanks for the pointer, I will verify it.

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:
```diff
@@ -824,46 +826,61 @@ public boolean deletePermanent(JobID jobId, PermanentBlobKey key) {
+        deleteLocalData(jobId);
+    }
 
+    private void deleteLocalData(JobID jobId) throws IOException {
```
do we need an extra method? can the globalCleanup simply call the localCleanup instead?

XComp replied:
We could do that. I just thought that it makes the reading of the code easier...

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/ThrowingJobGraphWriter.java:
```diff
@@ -31,8 +31,5 @@ public void putJobGraph(JobGraph jobGraph) {
-
-    @Override
-    public void releaseJobGraph(JobID jobId) {}
+    public void localCleanup(JobID jobId) {}
```
seems unnecessary

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerUtils.java:
```diff
@@ -44,15 +42,14 @@ private SchedulerUtils() {
             try {
-                return createCompletedCheckpointStore(
-                        configuration, userCodeLoader, checkpointRecoveryFactory, log, jobId);
+                return checkpointRecoveryFactory.createRecoveredCompletedCheckpointStore(
```
I'm not a big fan of moving this into the CheckpointRecoveryFactory interface. What is the intuition / benefit behind it?

XComp replied:
We need to instantiate the `CompletedCheckpointStore` in different locations (i.e. in `SchedulerBase` and in `CheckpointJobDataCleanupRunner`). I didn't want to duplicate the logic for accessing the configuration to get the number of retained checkpoints.

But instead, we could also move the static method from `SchedulerUtils` into some other utils class

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/JobGraphWriter.java:
```diff
@@ -42,5 +40,9 @@
      * @param jobId specifying the job to release the locks for
      * @throws Exception if the locks cannot be released
      */
```
Do we intentionally keep the old javadoc?

XComp replied:
Yes, I thought it would be more specific to the actual class.

XComp committed:
[FLINK-11813] Adds generic interfaces for cleaning up Job-related data

XComp committed:
[FLINK-11813] Refactors JobGraphWriter interface to implement LocallyCleanableResource and GloballyCleanableResource

XComp committed:
[FLINK-11813] Makes BlobServer implement LocallyCleanableResource and GloballyCleanableResource

XComp committed:
[FLINK-11813] Makes HighAvailabilityServices implement LocallyCleanableResource and GloballyCleanableResource

XComp committed:
[FLINK-11813] Makes JobManagerMetricGroup implement LocallyCleanableResource and GloballyCleanableResource

XComp committed:
[FLINK-11813] Adds ResourceCleaner interface and default implementation

XComp committed:
[FLINK-11813] Adds method to make the TestingJobManagerRunner fail during close

XComp committed:
[FLINK-11813] Adds JobManagerRunnerRegistry

XComp committed:
[FLINK-11813] Integrates JobManagerRunnerRegistry into Dispatcher

XComp committed:
[FLINK-11813] Refactors Dispatcher constructor signature

XComp committed:
[FLINK-11813] Moves JobManagerRunnerRegistry instantiation into separate constructor

XComp committed:
[FLINK-11813] Integrates ResourceCleaner into Dispatcher

XComp committed:
[FLINK-11813] Refactors createJobManagerRunner to be an initializing method

XComp committed:
[FLINK-11813] Integrates CheckpointResourcesCleanupRunner into Dispatcher

XComp committed:
[hotfix] Makes FileSystemBlobStore.(delete|deleteAll) comply to the BlobStore interface

Previously, FileSystemBlobStore's delete methods didn't return true if the
corresponding file didn't exist and, therefore, was not deleted. BlobStore.delete and
BlobStore.deleteAll have a different contract which states that, in that case, the
method should return true.

The internal FileSystemBlobStore.delete method got updated accordingly. Additionally, a
test class covering the FileSystemBlobStore was added

XComp commented:
Closed in favor of #4
